### DISCUSSION TITLE
How to List Full Model Topology, including Activation Layers

### DISCUSSION THREAD

**Discussion by oelghaza** [google/flax#4358] [Q&A]
I am looking for a way to reconstruct my NN model while using the new NNX API.  If I have a basic MLP with Activations between Linear layers, I am not able to retrieve any information on the Activation layers.

I understand this is due to them being JAX functions rather than NNX Modules.  Is there a way for me to perform model surgery and retrieve information about the activation layers of a saved model — without explicitly creating a Flax Module for each activation layer?

Ie if my model has the following architecture:
```
-Linear1
-Activation1
-Linear2
-Activation2
```

Using the model surgery techniques in the guides(`display()`,` iter_children()`, `iter_modules()`, saving/restoring checkpoint) I can only retrieve the following:
```
-Linear1
-Linear2
```

**Comment by cgarciae** [comment]
Hey! Flax NNX will print all instance attributes of a Module so if you refactor your code such that the activation functions are attributes you should be able to find them with `nnx.iter_graph`.

**Reply by oelghaza** [reply]
I tried this and I am able to see the attributes listed in the order they were created.  This does not show me the layer order for a forward pass of the network.

I am looking to view the actual layer-by-layer topology of the neural network.

**Reply by cgarciae** [reply]
Can you show an example of what is going wrong?

**Reply by oelghaza** [reply]
Hi, I just picked this back up. Thanks for helping!
I attached some screenshots to give a better idea.  

In this example I have a basic MLP that alternates between Dense and Activation Layers.  All the model surgery techniques I have tried list all the dense, followed by all the activation layers.  I am not able to view the actual topology which interleaves the activation layers between the dense ones; I also cannot find the activation as an attribute of the desne layers.  

I.e. I want to take in a model and see the topology of a forward pass:  [dense1, relu1, dense2, relu2, ...] instead of what I see [dense1, dense2, relu1, relu2].

Model def:
<img width="417" alt="image" src="https://github.com/user-attachments/assets/8bcd34c7-0540-4ac0-a233-b4834ab263b1" />


iter_graph() output:
<img width="261" alt="image" src="https://github.com/user-attachments/assets/fb30cab6-4a1c-4a5f-ba30-84f09748f589" />


graphdef output:
<img width="1122" alt="image" src="https://github.com/user-attachments/assets/c981839a-9466-4ada-90c7-a4bdee405c8e" />

**Reply by pablordoricaw** [reply]
Hey @oelghaza, here's a way to implement a neural network to see the anatomy of a forward pass when you look at/traverse the model.

@cgarciae -  I'd appreciate your thoughts on this approach.

The idea is to leverage the [nnx.Sequential](https://flax.readthedocs.io/en/latest/_modules/flax/nnx/helpers.html#Sequential) class in combination with your `relu_module` class, and an additional `softmax_module` class:

```python
from flax import nnx

class relu_module(nnx.Module):
    def __call__(self, x):
        return nnx.relu(x)

class softmax_module(nnx.Module):
    def __call__(self, x):
        return nnx.softmax(x)
        
nnx_1_jet_mlp = nnx.Sequential(
    nnx.Linear(64, 32, rngs=nnx.Rngs(0)),
    relu_module(),

    nnx.Linear(64, 32, rngs=nnx.Rngs(0)),
    relu_module(),

    nnx.Linear(64, 32, rngs=nnx.Rngs(0)),
    relu_module(),

    nnx.Linear(64, 32, rngs=nnx.Rngs(0)),
    softmax_module()
)
```

If you create your model as an instance of the `nnx.Sequential` class, then the `GraphDef` for the `nnx_1_jet_mlp` model has the layers sorted in the order of the forward pass:

![image](https://github.com/user-attachments/assets/e18a4684-40ca-4519-aaa9-b6aafa9e7bec)

Traversing it with the `iter_modules()` method:

![image](https://github.com/user-attachments/assets/63db2543-1597-42be-9097-8e65844a5f88)
